### DISCUSSION TITLE
Passing "dependencies" to route handlers, & performance implications

### DISCUSSION THREAD

**Discussion by castarco** [fastify/fastify#3698] [Q&A] [answered]
Hi, I'm writting a small internal "service template" built on top of Fastify (basically a highly opinionated mini-framework). One of the things I did there is trying to clearly distinguish between infrastructure, application & domain code (a bit of DDD, you might now the drill).

I know that some parts of the code I'll show here might cause disgust to performance experts 😅 , but that's part of the reason why I'm writing here, so please bear with me.

One of the techniques I apply is to rely on inversion of control, through dependencies injection. Besides that, I also made the code to enforce that all "controllers" implement proper validation logic. The easiest way for a guy like me was to implement part of this through classes. Here some relevant parts of the code:

Controller interface I use:
```typescript
export type RequestHandler<
  RequestType extends Partial<RouteGenericInterface> = Record<string, never>,
  This extends HttpController<RequestType> = HttpController<RequestType>,
> = (
  this: This,
  request: GenericRequest<RequestType>,
  reply: GenericResponse,
  fastify: FastifyApp,
) => Promise<void>

export interface HttpController<
  RequestType extends Partial<RouteGenericInterface> = Record<string, never>,
> {
  readonly method: HttpMethod | HttpMethod[]
  readonly path: string

  handler: RequestHandler<RequestType, HttpController<RequestType>>

  getHandler(): FastifyRequestHandler<RequestType>
  getSchemas(): ControllerSchemas
}
```

Abstract controller class I use. This is one of the spots I believe that might be breaking most of the performance benefits from Fastify, and actually I copied some of the comments verbatim, precisely talking about this point:

```typescript
/**
 * This abstract class exists with the main goal of being able to keep two
 * contexts at the same time inside handler functions (one for the controller,
 * and another one for the Fastify app).
 *
 * i.e.: Inside the `handler` method, this will point to the controller object,
 *       and `fastify` will obviously point to the Fastify app.
 *
 * Keeping a reference to the controller object might be useful when we want to
 * keep a reference to "useCase" factories, among other dependencies.
 */
export abstract class BaseController<RequestType>
  implements HttpController<RequestType>
{
  abstract method: HttpMethod | HttpMethod[]
  abstract path: string

  abstract handler(
    request: GenericRequest<RequestType>,
    reply: GenericResponse,
    fastify: FastifyApp,
  ): Promise<void>

  // TODO: Instead of wrapping the handler to pass our controller as `this`,
  //       consider using
  //       `https://www.fastify.io/docs/latest/Reference/Hooks/#prehandler`
  //       passing a handler function that reads from a "shared" table to take
  //       a controller reference and use it to "decorate" the request.
  //       Every time we "install" a route, register the controller into that
  //       table.
  //       As to why I consider this approach, my reason is that, eventually,
  //       we might end up using that hook anyway, and we would pay the price
  //       of calling a wrapper function twice. In any case, we can wait until
  //       then.
  getHandler(): FastifyRequestHandler<RequestType> {
    // Here, `this` still referes to our controller
    const controller: BaseController<RequestType> = this

    // Not using an arrow function here so we can define `this` as its first
    // parameter, and also specify its type as `FastifyApp`.
    return async function (
      this: FastifyApp,
      request: GenericRequest<RequestType>,
      reply: GenericResponse,
    ): Promise<void> {
      // But here, `this` refers to the Fastify app
      await controller.handler(request, reply, this)
    }
  }

  abstract getSchemas(): ControllerSchemas
}
```

This is an example of a dummy controller, implemented on top of base controller (complete file):

```typescript
import { GenericRequest, GenericResponse } from '../core/HttpController'
import { JsonType, RequestGenericInterface } from '../core/ControllerSchemas'
import { BaseController } from '../core/BaseController'

// We'll pass this schema to Fastify to validate our requests
const controllerSchema = {
  query: JsonType.Object({ name: JsonType.String() }),
}

// This type is to know what we can do with the request from TypeScript's perspective
type RequestShape = RequestGenericInterface<typeof controllerSchema>

export class GreetController extends BaseController<RequestShape> {
  // "metadata" that will be used to configure the endpoint associated to this controller.
  // it could be "static" (associated to the class, instead of the instance), but it gets
  // annoying pretty fast when coding in TypeScript.
  public readonly method = 'GET'
  public readonly path = '/greet'

  public async handler(
    request: GenericRequest<RequestShape>,
    reply: GenericResponse,
  ): Promise<void> {
    const name = request.query.name
    await reply.send({ hello: name })
  }

  // This one could be static too, same: too annoying
  public getSchemas() {
    return controllerSchema
  }
}
```

And this is how I "install" my controllers into specific routes:

```typescript
// That's why I have an interface for controller objects in the first place, so I can easily
// pass them to this function, and "configure" the routes. I also do it to enforce that every
// controller always provides a strict schema to allow Fastify&AJV to validate requests.

function installRoute<RequestType>(
  app: FastifyApp,
  controller: HttpController<RequestType>,
): void {
  app.route({
    method: controller.method,
    url: controller.path,
    handler: controller.getHandler(),
    schema: controller.getSchemas(),
  })
}
```

Careful observation shows that some methods don't need to be "instance methods", they could easily be static ones, and therefore they could be extracted into "free" functions (not tied to any class or object). I use instance methods just for convenience, they are called just once, and it's just easier to define the types and encapsulate some stuff.

Also, from this example, it seems obvious that we don't need so much fuzz, the controllers seem to be very simple functions, and they don't need to be encapsulated into objects. But that's just for this particular example, in my "real" cases, I inject "use case" factories that depend on a whole bunch of other dependencies (repositories, ORM, etc).

But here is the thing, it feels... strange, having to go through the contortion I apply in `BaseController` so I can pass a reference to the controller itself to the handler function. And I fear I might be killing performance too there. I still didn't had time to run a profiler to confirm my suspicions, but I think some people could have more immediate & insightful intuitions about it.

Also, I was hoping that maybe someone has good examples in mind I could check for, when it comes to the problem of injecting dependencies to route handlers. I know I can rely on some of Fastify's hooks, but it's not immediately clear what would be the best combination for this kind of code.

Thank you for your attention!

**Comment by mcollina** [comment, marked as answer]
The simple mechanism that always worked for me is the fastify plugin system and decorators. Note that we even have a dependency check system in https://github.com/fastify/fastify/blob/6363ff4eae89cd38dc8b901e5f9c23ce4d57bc0a/lib/pluginUtils.js#L55, but I didn't find it documented 🤦‍♂️.


If you want something a bit automatic, fastify-autoload should provide that, there are plenty of examples in the test folder.

Overall this is something we should document better.

**Reply by castarco** [reply]
Thx 🙂 🙇🏻 , I'll take a look into this.
